### PR TITLE
Fixed #27533 -- Fixed inspectdb crashing on unique index with Coalesce.

### PR DESCRIPTION
**Pull request by weima**
https://code.djangoproject.com/ticket/27533

### CODE REVIEW

**Comment by timgraham** [comment]
Could you elaborate on the steps to reproduce the issue and add a regression test if possible?

**Comment by weima** [comment]
We had a unique constraint created with 
 CREATE UNIQUE INDEX ux_my_table
    ON my_table (id1, id2, COALESCE(id3, -1));

id1, id2, and id3 are all foreign keys. id1 and id2 are not nullable, but id3 is. The command line crashes at this table. 


I am not sure how to add a regression test for this.

**Comment by timgraham** [comment]
I'm not sure if Django should output something for that unique constraint since what Django generates won't be equivalent. What behavior does your PR give? For a test, look in `tests/inspectdb`. You might have to make the test PostgreSQL specific and run a raw SQL creator to create that index.

**Comment by adamchainz** [comment]
PR title should have two dashes, not three

**Comment by weima** [comment]
The result after applying my fix generated a model class 

class MyTableModel(models.Model):
    id1 = models.ForeignKey(FT1, models.DO_NOTHING)
    id2 = models.ForeignKey(FT2, models.DO_NOTHING)
    id3 = models.ForeignKey(FT3, models.DO_NOTHING, blank=True, null=True)

    class Meta:
        managed = False
        db_table = 'my_table'
        **unique_together = (('id1', 'id2'),)**

So yes, the generated table is not the same as the database table. But It at least let me finish the process and I can come back to fix the code manually. 

I don't know if this is the best we can do, or if it's even a viable temporary solution. 

I will create the tests if you think it's worth continuing with this direction. Otherwise I will just keep this fix on my fork. It does fix my problem, but may not be the best idea for the project as whole.

**Comment by timgraham** [comment]
A solution that prevents the crash while also adding a comment in the models file that the unique constraint contains some other data that Django couldn't introspect could be ideal for now unless you wanted to pursue the more ambitious task of supporting expressions like `Coalesce` in `unique_together`.

**Comment by weima** [comment]
Cool, I will continue working on it then. Do you have a code example of adding comment ? I will need to do that.

**Comment by timgraham** [comment]
Look at the `comment_notes` variable in this file for a similar thing.

**Comment by weima** [comment]
OK, I added a comment when detecting unsupported constraints, still not sure how to add a test. Is it absolutely necessary to add a test for this case?

**Comment by timgraham** [comment]
Yes, we don't commit changes without tests unless it's infeasible or impractical which doesn't seem to be the case here.

File: django/core/management/commands/inspectdb.py [hunk 1]
```diff
@@ -270,18 +270,29 @@ def get_meta(self, table_name, constraints, column_to_field_name):
...
-        meta = ["",
-                "    class Meta:",
-                "        managed = False",
-                "        db_table = '%s'" % table_name]
+
+        meta = [""]
+
+        if has_unsupported_constraint_columns:
+            unsupported_warning_comments = ["    # " + table_name + " contains unsupported constraints. ",
+                                            "    # Please modify the model class to match database constraint"]
```
Comment (timgraham) [inline django/core/management/commands/inspectdb.py]: I'm not sure about "Please modify..." considering that Django doesn't support constraints like the one you encountered.

**Comment by weima** [comment]
@timgraham I have changed the wording for comments.  As I am not familiar with the testing part. Is there away to team up with someone to finish it? Maybe someone can fork my branch and add the testing there?

**Comment by timgraham** [comment]
Yes, if you can't add tests, then close your PR and ask for help on the ticket.

**Comment by weima** [comment]
Won't able to add a test. Will ask for help on the ticket instead.

**Pull request closed without merging**